src/JustEat.StatsD/PooledUdpTransport.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Net.Sockets;
+using System.Text;
+using JustEat.StatsD.EndpointLookups;
+
+namespace JustEat.StatsD
+{
+    /// <summary>
+    /// A class representing an implementation of <see cref="IStatsDTransport"/> uses UDP and pools sockets. This class cannot be inherited.
+    /// </summary>
+    public sealed class PooledUdpTransport : IStatsDTransport, IDisposable
+    {
+        private const int PoolSize = 30;
+
+        private readonly SimpleObjectPool<Socket> _socketPool
+            = new SimpleObjectPool<Socket>(PoolSize, pool => UdpTransport.CreateSocket());
+
+        private readonly IPEndPointSource _endpointSource;
+        private bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PooledUdpTransport"/> class.
+        /// </summary>
+        /// <param name="endPointSource">The <see cref="IPEndPointSource"/> to use.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="endPointSource"/> is <see langword="null"/>.
+        /// </exception>
+        public PooledUdpTransport(IPEndPointSource endPointSource)
+        {
+            _endpointSource = endPointSource ?? throw new ArgumentNullException(nameof(endPointSource));
+        }
+
+        /// <summary>
+        /// Finalizes an instance of the <see cref="PooledUdpTransport"/> class.
+        /// </summary>
+        ~PooledUdpTransport()
+        {
+            Dispose(false);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <inheritdoc />
+        public void Send(string metric)
+        {
+            if (string.IsNullOrWhiteSpace(metric))
+            {
+                return;
+            }
+
+            var bytes = Encoding.UTF8.GetBytes(metric);
+            var endpoint = _endpointSource.GetEndpoint();
+
+            var socket = _socketPool.Pop();
+
+            if (socket == null)
+            {
+                return;
+            }
+
+            try
+            {
+                socket.SendTo(bytes, endpoint);
+            }
+            finally
+            {
+                _socketPool.Push(socket);
```

Comment:
If the socket has failed, I don't think it should go back into the pool.

Reply:
True, but with the existing pool implementation this would slowly starve the pool until there were no Sockets available and then no stats would be emitted at all.